src/main/java/com/zendesk/maxwell/schema/SchemaPosition.java

```diff
@@ -36,39 +38,46 @@ public SchemaPosition(ConnectionPool pool, Long serverID) {
 
 	public void start() {
 		this.thread = new Thread(this, "Position Flush Thread");
-		this.run.set(true);
 		thread.start();
 	}
 
-	public void stop() {
-		this.run.set(false);
-
+	public void stopLoop() throws TimeoutException {
+		this.requestStop();
 		thread.interrupt();
+		super.stopLoop();
+	}
 
-		while ( thread.isAlive() ) {
-			try {
-				Thread.sleep(10);
-			} catch (InterruptedException e) { }
+	@Override
+	protected void beforeStop() {
+		if ( exception == null ) {
+			LOGGER.info("Storing final position: " + position.get());
+			store(position.get());
 		}
 	}
 
 	@Override
 	public void run() {
-		while ( run.get() && this.exception == null ) {
-			BinlogPosition newPosition = position.get();
+		try {
+			runLoop();
+		} catch ( Exception e ) {
+			// this code should never be hit.  There is, I suppose,
+			// a design flaw in my code, but at a certain point
+			// the whole inheritance + exception handling thing
+			// just started to drive me nuts.
+			LOGGER.error("Hit unexpected exception in SchemaPosition thread: " + e);
```

Comment:
sorry for the long rant.  I don't think I'm really much of a fan of checked exceptions.